File: server/routes/estimates.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from database import db
import json

estimates_bp = Blueprint('estimates', __name__)

@estimates_bp.route('/', methods=['POST'])
@jwt_required()
def save_estimate():
    """Save a new estimate"""
    try:
        user_id = get_jwt_identity()
        data = request.get_json()
        
        project_name = data.get('project_name', 'Untitled Project')
        plot_length = data.get('plot_length')
        plot_breadth = data.get('plot_breadth')
        total_area = data.get('total_area')
        num_floors = data.get('num_floors')
        material_quality = data.get('material_quality', 'standard')
        total_cost = data.get('total_cost')
        cost_per_sqft = data.get('cost_per_sqft')
        estimate_data = data.get('estimate_data', {})
        
        if not all([plot_length, plot_breadth, total_area, num_floors, total_cost]):
            return jsonify({'error': 'Missing required fields'}), 400
        
        estimate_id = db.execute_query(
            """INSERT INTO estimates 
            (user_id, project_name, plot_length, plot_breadth, total_area, 
            num_floors, material_quality, total_cost, cost_per_sqft, estimate_data)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)""",
            (user_id, project_name, plot_length, plot_breadth, total_area,
             num_floors, material_quality, total_cost, cost_per_sqft, 
             json.dumps(estimate_data))
        )
        
        if estimate_id:
            return jsonify({
                'message': 'Estimate saved successfully',
                'estimate_id': estimate_id
            }), 201
        else:
            return jsonify({'error': 'Failed to save estimate'}), 500
            
    except Exception as e:
        print(f"Save estimate error: {e}")
        return jsonify({'error': 'Internal server error'}), 500

@estimates_bp.route('/', methods=['GET'])
@jwt_required()
def get_estimates():
    """Get all estimates for the current user"""
    try:
        user_id = get_jwt_identity()
        
        estimates = db.execute_query(
            """SELECT id, project_name, plot_length, plot_breadth, total_area,
            num_floors, material_quality, total_cost, cost_per_sqft, created_at
            FROM estimates WHERE user_id = %s ORDER BY created_at DESC""",
            (user_id,),
            fetch=True
        )
        
        return jsonify({'estimates': estimates or []}), 200
        
    except Exception as e:
        print(f"Get estimates error: {e}")
        return jsonify({'error': 'Internal server error'}), 500

@estimates_bp.route('/<int:estimate_id>', methods=['GET'])
@jwt_required()
def get_estimate(estimate_id):
    """Get a specific estimate"""
    try:
        user_id = get_jwt_identity()
        
        estimate = db.execute_query(
            "SELECT * FROM estimates WHERE id = %s AND user_id = %s",
            (estimate_id, user_id),
            fetch_one=True
        )
        
        if not estimate:
            return jsonify({'error': 'Estimate not found'}), 404
        
        # Parse JSON data
        if estimate.get('estimate_data'):
            estimate['estimate_data'] = json.loads(estimate['estimate_data'])
        
        return jsonify({'estimate': estimate}), 200
        
    except Exception as e:
        print(f"Get estimate error: {e}")
        return jsonify({'error': 'Internal server error'}), 500

@estimates_bp.route('/<int:estimate_id>', methods=['DELETE'])
@jwt_required()
def delete_estimate(estimate_id):
    """Delete an estimate"""
    try:
        user_id = get_jwt_identity()
        
        # Check if estimate exists and belongs to user
        estimate = db.execute_query(
            "SELECT id FROM estimates WHERE id = %s AND user_id = %s",
            (estimate_id, user_id),
            fetch_one=True
        )
        
        if not estimate:
            return jsonify({'error': 'Estimate not found'}), 404
        
        db.execute_query(
            "DELETE FROM estimates WHERE id = %s",
            (estimate_id,)
        )
        
        return jsonify({'message': 'Estimate deleted successfully'}), 200
        
    except Exception as e:
        print(f"Delete estimate error: {e}")
        return jsonify({'error': 'Internal server error'}), 500

@estimates_bp.route('/<int:estimate_id>', methods=['PUT'])
@jwt_required()
def update_estimate(estimate_id):
    """Update an estimate"""
    try:
        user_id = get_jwt_identity()
        data = request.get_json()
        
        # Check if estimate exists and belongs to user
        estimate = db.execute_query(
            "SELECT id FROM estimates WHERE id = %s AND user_id = %s",
            (estimate_id, user_id),
            fetch_one=True
        )
        
        if not estimate:
            return jsonify({'error': 'Estimate not found'}), 404
        
        # Build update query
        updates = []
        params = []
        
        if 'project_name' in data:
            updates.append("project_name = %s")
            params.append(data['project_name'])
        
        if 'estimate_data' in data:
            updates.append("estimate_data = %s")
            params.append(json.dumps(data['estimate_data']))
        
        if not updates:
            return jsonify({'error': 'No fields to update'}), 400
        
        params.append(estimate_id)
        query = f"UPDATE estimates SET {', '.join(updates)} WHERE id = %s"
        
        db.execute_query(query, tuple(params))
        
        return jsonify({'message': 'Estimate updated successfully'}), 200
        
    except Exception as e:
        print(f"Update estimate error: {e}")
        return jsonify({'error': 'Internal server error'}), 500

File: public/js/estimates.js
/**
 * Estimate Management Module
 * Handles saving, loading, and managing construction estimates
 */

// Current estimate data
let currentEstimate = null;

// ===================================
// SAVE ESTIMATE
// ===================================

async function saveEstimate(estimateData) {
    if (!api.isAuthenticated()) {
        showToast('Please login to save estimates');
        openModal('login-modal');
        return;
    }

    try {
        const response = await api.saveEstimate(estimateData);
        showToast('Estimate saved successfully!');
        currentEstimate = response.estimate;
        return response;
    } catch (error) {
        showToast(error.message || 'Failed to save estimate');
        throw error;
    }
}

// ===================================
// LOAD ESTIMATES
// ===================================

async function loadEstimates() {
    if (!api.isAuthenticated()) {
        showToast('Please login to view estimates');
        return [];
    }

    try {
        const response = await api.getEstimates();
        return response.estimates || [];
    } catch (error) {
        showToast('Failed to load estimates');
        return [];
    }
}

// ===================================
// DELETE ESTIMATE
// ===================================

async function deleteEstimate(estimateId) {
    if (!confirm('Are you sure you want to delete this estimate?')) {
        return;
    }

    try {
        await api.deleteEstimate(estimateId);
        showToast('Estimate deleted successfully');
        await showEstimatesModal();
    } catch (error) {
        showToast(error.message || 'Failed to delete estimate');
    }
}

// ===================================
// SHOW ESTIMATES MODAL
// ===================================

async function showEstimatesModal() {
    const estimates = await loadEstimates();
    const modal = document.getElementById('estimates-modal');
    const list = document.getElementById('estimates-list');

    if (estimates.length === 0) {
        list.innerHTML = `
            <div style="text-align: center; padding: 40px; color: var(--text-secondary);">
                <i class="fas fa-folder-open" style="font-size: 3rem; margin-bottom: 20px; opacity: 0.5;"></i>
                <p>No saved estimates yet</p>
                <p style="font-size: 0.875rem;">Use any calculator and click "Save Estimate" to get started</p>
            </div>
        `;
    } else {
        list.innerHTML = estimates.map(estimate => `
            <div class="estimate-item">
                <div class="estimate-header">
                    <h4>${estimate.project_name || 'Untitled Project'}</h4>
                    <span class="estimate-date">${new Date(estimate.created_at).toLocaleDateString('en-IN')}</span>
                </div>
                <div class="estimate-details">
                    <div class="estimate-detail">
                        <i class="fas fa-ruler-combined"></i>
                        <span>${formatNumber(estimate.total_area, 0)} sq.ft</span>
                    </div>
                    <div class="estimate-detail">
                        <i class="fas fa-building"></i>
                        <span>${estimate.num_floors} floor(s)</span>
                    </div>
                    <div class="estimate-detail">
                        <i class="fas fa-star"></i>
                        <span>${estimate.material_quality}</span>
                    </div>
                </div>
                <div class="estimate-cost">
                    <span class="cost-label">Total Cost:</span>
                    <span class="cost-value">${formatCurrency(estimate.total_cost)}</span>
                    <span class="cost-per-sqft">(${formatCurrency(estimate.cost_per_sqft)}/sq.ft)</span>
                </div>
                <div class="estimate-actions">
                    <button class="btn btn-sm btn-secondary" onclick="loadEstimateDetails(${estimate.id})">
                        <i class="fas fa-eye"></i> View
                    </button>
                    <button class="btn btn-sm btn-primary" onclick="deleteEstimate(${estimate.id})">
                        <i class="fas fa-trash"></i> Delete
                    </button>
                </div>
            </div>
        `).join('');
    }

    openModal('estimates-modal');
}

// ===================================
// LOAD ESTIMATE DETAILS
// ===================================

async function loadEstimateDetails(estimateId) {
    try {
        const response = await api.getEstimate(estimateId);
        const estimate = response.estimate;

        // Display estimate details
        alert(`
Project: ${estimate.project_name}
Area: ${estimate.total_area} sq.ft
Floors: ${estimate.num_floors}
Quality: ${estimate.material_quality}
Total Cost: ${formatCurrency(estimate.total_cost)}
Cost per sq.ft: ${formatCurrency(estimate.cost_per_sqft)}
        `);

        // TODO: Populate calculator with estimate data
    } catch (error) {
        showToast('Failed to load estimate details');
    }
}

// ===================================
// SAVE FROM CALCULATOR
// ===================================

async function saveFromCalculator(calculatorType, data) {
    const projectName = prompt('Enter project name:');
    if (!projectName) return;

    const estimateData = {
        project_name: projectName,
        plot_length: data.length || 0,
        plot_breadth: data.breadth || 0,
        total_area: data.total_area || 0,
        num_floors: data.num_floors || 1,
        material_quality: data.quality || 'standard',
        total_cost: data.total_cost || 0,
        cost_per_sqft: data.cost_per_sqft || 0,
        estimate_data: JSON.stringify(data)
    };

    await saveEstimate(estimateData);
}

// Make functions globally available
window.saveEstimate = saveEstimate;
window.loadEstimates = loadEstimates;
window.deleteEstimate = deleteEstimate;
window.showEstimatesModal = showEstimatesModal;
window.loadEstimateDetails = loadEstimateDetails;
window.saveFromCalculator = saveFromCalculator;
